File: tools/blobinator.c
#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <stdlib.h>
#include <fts.h>

#define VERSION "1.0"

#define TYPE_TEXTURE() strcmp(type, "texture") == 0

struct buffer {
    char* data;
    size_t size, alloc;
};

struct data_entry {
    const char* data_name;
    char* start, datastart, end;
};

struct texture_header {
    char* texture_name;
    unsigned short width, height;
    unsigned char format, size;
};

char* template[] = {
    "#include <%s.h>\n\n",
    "const size_t %s = %zu\n\n",
    "const tex_info %s[] = {\n",
    "   %s\n",
    "};"
};

char* input_path, output_path, c_path;
char* type, extension;
char* texmap_path;

struct texture_header* texheaders;

char* get_string_from_file(FILE* file) {
    char* buffer = 0;
    size_t length;
    
    fseek(file, 0, SEEK_END);
    length = ftell(file);
    fseek(file, 0, SEEK_SET);
    buffer = malloc(length);
    fread(buffer, 1, length, file);

    return buffer;
}

size_t create_texture_header_array() {
    FILE* map = fopen(texmap_path, "r");
    size_t entries = 0;
    char c;
    
    if(map == NULL) {
        printf("Unable to open Texture Metadata Map at \"%s\"!", texmap_path);
        exit(1);
    }

    for(c = getc(map); c != EOF; c = getc(map))
        if(c == '\n')
            entries++;
    
    struct texture_header* headers[entries];
    char* mapdata = get_string_from_file(map);
    fclose(map);

    char* line = strok(mapdata, '\n');
    for(size_t i = 0; line != NULL; i++) {
        char* metadata[3];
        struct texture_header* header;
    
        char* ptr = strtok(line, ' ');
        for(size_t i = 0; ptr != NULL; i++) {
            metadata[i] = ptr;
            ptr = strtok(NULL, ' ');
        }
    
        header->texture_name = metadata[0];
        header->format = strok(metadata[1], '/');
        header->size = strtok(NULL, '/');
        header->width = strtok(metadata[2], 'x');
        header->height = strtok(NULL, 'x');

        headers[i] = header;
        line = strok(NULL, '\n');
    }

    texheaders = headers;
    return entries;
}

struct texture_header* get_texture_header(char* path) {
    for(size_t i = 0; i < sizeof(texheaders); i++)
        if(strcmp(texheaders[i].texture_name, path) == 0)
            return &texheaders[i];
}

void append_data(struct buffer *buf, void *data, size_t size) {
    if (size > buf->alloc - buf->size) {
        size_t new_alloc = buf->alloc;
        if (new_alloc == 0) {
            new_alloc = 64;
        }
        while (size > new_alloc - buf->size) {
            new_alloc *= 2;
        }
        char *new_data = realloc(buf->data, new_alloc);
        if (new_data == NULL) {
            printf("Out of memory!");
            exit(1);
        }
        buf->data = new_data;
        buf->alloc = new_alloc;
    }
    memcpy(buf->data + buf->size, data, size);
    buf->size += size;
}

int endsWith(const char* str, const char* suffix) {
    if(!str || !suffix)
        return 0;
    size_t lenstr = strlen(str);
    size_t lensuffix = strlen(suffix);
    if (lensuffix >  lenstr)
        return 0;
    return strncmp(str + lenstr - lensuffix, suffix, lensuffix) == 0;
}

void remove_substring(char* string, const char* substring) {

    while(string = strstr(string, substring))
        memmove(string, string + strlen(substring), 1 + strlen(string + strlen(substring)));
}

void print_version() {
    printf("Blobinator %s by Nicolas \"Vatuu\" Adamoglou\n", VERSION);
    printf("  A tool to package data assets and generate a identifier-based\n");
    printf("  C file for accessing data without static definitions.\n\n");
}

void print_usage() {
    print_version();
    printf("Options:\n");
    printf("  -i <Include Directory Path>\n");
    printf("     Directory containing the files to be included.\n");
    printf("  -t [TEXTURE|MODEL|SOUND]\n");
    printf("     Defines the data type to be blobbed.\n     Will be used by the generated C template.\n");
    printf("  -o <Path>\n");
    printf("     Path at which the blobfish will be generated.\n");
    printf("  -e <Extension>\n");
    printf("     Extension of the files to be blobbed.\n");
    printf("  -m\n");
    printf("     Path at which the C file will be generated.\n");
    printf("  -d\n");
    printf("     Path to the texture metadata map (TEXTURES only).\n");
    printf("  -v\n");
    printf("     Prints the version and description of the Blobinator.\n");
    printf("  -h -?\n");
    printf("     Prints this message.\n");
}

int main(int argc, char* argv[]) {
    int option;
    while((option = getopt(argc, argv, "i:t:e:o:v?h")) != -1) {
        switch(option) {
            case 'i':
                input_path = optarg;
                break;
            case 't':
                type = optarg;
                break;
            case 'm':
                c_path = optarg;
                break;
            case 'd':
                texmap_path = optarg;
                break;
            case 'o':
                output_path = optarg;
                break;
            case 'e': 
                extension = optarg;
                break;
            case 'v':
                print_version();
                return 0;
            case '?':
            case 'h':
            default:
                print_usage();
                return 0;
        }
    }



    char* dirs[2] = {input_path, NULL};
    FTS* ftsp = fts_open(dirs, FTS_LOGICAL, NULL);
    if(ftsp == NULL) {
        printf("Unable to open directory \"%s\"!", input_path);
        return 1;
    }

    size_t count = 0;
    struct buffer buf = {};


    if(TYPE_TEXTURE()) 
        count = create_texture_header_array();
    
    char* centries = "";

    FTSENT* ent;
    for(size_t i = 0; (ent = fts_read(ftsp)) != NULL && i < count; i++) {
        char* path = ent->fts_accpath;
        if(endsWith(path, extension)) {

            size_t start = buf.size;

            if(TYPE_TEXTURE()) {
                struct texture_header* header = get_texture_header(path);
                append_data(&buf, header->format, sizeof(unsigned char));
                append_data(&buf, header->size, sizeof(unsigned char));
                append_data(&buf, header->width, sizeof(unsigned short));
                append_data(&buf, header->height, sizeof(unsigned short));
            }

            count++;
            size_t size, headersize;
            FILE *data = fopen(path, "rb");

            fseek(data, 0, SEEK_END);
            size = ftell(data);
            fseek(data, 0, SEEK_SET);

            char* tmpBuf[32 * 1024];
            while (1) {
                size_t r = fread(tmpBuf, 1, sizeof(tmpBuf), data);
                if (r == 0)
                    break;
                append_data(&buf, tmpBuf, r);
            }
            
            if((size & 1)) {
                append_data(&buf, 0x00, 1);
                size++;
            }

            if(TYPE_TEXTURE())
                size += 6;

            size_t datastart = start + (TYPE_TEXTURE() ? 6 : 0);
            size_t dataend = buf.size;

            remove_substring(path, input_path); 
            const char* formatted;
            sprintf(formatted, "{\"%s\", %zu, %zu, %zu},\n", strok(path, '.'), start, datastart, dataend);
            strcat(centries, formatted);

            fclose(data);
        }
    }

    fts_close(ftsp);

    FILE* output_blob = fopen(output_path, "wb");
    if(output_blob == NULL) {
        printf("Unable to create output blob file \"%s\"", output_path);
        exit(1);
    }
    fwrite(buf.data, 1, buf.size, output_blob);
    fclose(output_blob);

    if(TYPE_TEXTURE()) {
        sprintf(template[0], "textures");
        sprintf(template[1], "TEXTURE_COUNT", count);
        sprintf(template[2], "TEXTURES");
        sprintf(template[3], centries);
    }

    FILE* output_c = fopen(c_path, "w+");
        if(output_blob == NULL) {
        printf("Unable to create output c file \"%s\"", output_path);
        exit(1);
    }

    fputs(template[0], output_blob);
    fputs(template[1], output_blob);
    fputs(template[2], output_blob);
    fputs(template[3], output_blob);
    fputs(template[4], output_blob);

    fclose(output_c);

    printf("DONE!");

    return 0;
}

File: tools/blobinator.cs
using System.Net;
using System.Collections;
using System.ComponentModel;
using System.Diagnostics;
//Yes

using System.IO;
using System;
using System.Collections.Generic;

namespace Reality.Tools {
    class Blobinator {

        private enum TextureFormat {
            RGBA = 0,
            YUV = 1,
            CI = 2,
            IA = 3,
            I = 4
        }
    
        private struct Options {
            public string input_dir, input_meta;
            public string output_path, output_path_map;
            public string extension;
            public Type type;
        }

        private enum Type {
            TEXTURES,
            MODELS,
            SOUND,
            MUSIC
        }

        private struct DataEntry {
            public string data_id;
            public uint start, dataStart, end;
        }

        private struct TextureMetadata {
            public string fileName;
            public ushort width, height;
            public byte format, size;

            public void populate(string[] meta, string id) {
                fileName = id;
                string[] formatString = meta[1].Split('/');
                format = (byte)(TextureFormat)Enum.Parse(typeof(TextureFormat), formatString[0].ToUpper());
                size = (byte)(Math.Log(int.Parse(formatString[1]), 2) - 2);
                string[] sizeString = meta[2].Split('x');
                byte[] tw = BitConverter.GetBytes(ushort.Parse(sizeString[0]));
                Array.Reverse(tw);
                width = BitConverter.ToUInt16(tw, 0);
                byte[] th = BitConverter.GetBytes(ushort.Parse(sizeString[1]));
                Array.Reverse(th);
                height = BitConverter.ToUInt16(th, 0);
            }

            public byte[] createData() {
                using(MemoryStream stream = new MemoryStream()) {
                    using(BinaryWriter writer = new BinaryWriter(stream)) {
                        writer.Write(width); writer.Write(height);
                        writer.Write(format); writer.Write(size);
                        writer.Write((ushort)0xFF);
                        return stream.ToArray();
                    }
                }
            }
        }

        static void Main(string[] args) {
            var options = ParseOptions(args);

            if(options.type == Type.TEXTURES) {
                Console.WriteLine("\nData Mode: TEXTURES\n");
                Dictionary<string, TextureMetadata> data = GetTextureData(options);
                foreach(KeyValuePair<string, TextureMetadata> entry in data) {
                    Console.WriteLine("DataNode: {0}", entry.Key);
                    Console.WriteLine("  {0} {1}/{2} {3}x{4}", entry.Value.fileName, entry.Value.format, entry.Value.size, entry.Value.width, entry.Value.height);
                }
                List<DataEntry> metadata = CreateBlobDataTexture(data, options.output_path);
                CreateLookupTableTexture(metadata, options.output_path_map);
            }

            Console.WriteLine("Done!");
        }

        private static Options ParseOptions(string[] args) {
            Options option = new Options();

            for (int i = 0; i < args.Length; i++)
            {
                string arg = args[i];

                switch(arg) {
                    case "-i":
                    case "-input":
                        option.input_dir = args[++i];
                        break;
                    case "-d":
                    case "-data":
                        option.input_meta = args[++i];
                        break;
                    case "-o":
                    case "-output":
                        option.output_path = args[++i];
                        break;
                    case "-m":
                    case "-map":
                        option.output_path_map = args[++i];
                        break;
                    case "-t":
                    case "-type":
                        option.type = (Type)Enum.Parse(typeof(Type), args[++i]);
                        break;
                    case "-e":
                    case "-extension":
                        option.extension = args[++i];
                        break;
                    case "-h":
                    case "-?":
                    case "-help":
                    case "--help":
                        PrintUsage();
                        Environment.Exit(0);
                        break;
                    default:
                        Console.WriteLine("Unknown Option: " + arg);
                        break;
                        
                }
            }

            return option;
        }

        private static void PrintUsage() {
            Console.WriteLine("Blobinator {0} by Nicolas \"Vatuu\" Adamoglou", "1.0");
            Console.WriteLine("  A tool to package data assets and generate a identifier-based");
            Console.WriteLine("  C file for accessing data without static definitions.\n");
            Console.WriteLine("Options:");
            Console.WriteLine("  -i -include");
            Console.WriteLine("     Directory containing the files to be included.");
            Console.WriteLine("  -t -type [TEXTURE|MODEL|SOUND]");
            Console.WriteLine("     Defines the data type to be blobbed.\n     Will be used by the generated C template.");
            Console.WriteLine("  -o -output");
            Console.WriteLine("     Path at which the blobfish will be generated.");
            Console.WriteLine("  -e -extension");
            Console.WriteLine("     Extension of the files to be blobbed.");
            Console.WriteLine("  -m -map");
            Console.WriteLine("     Path at which the C file will be generated.");
            Console.WriteLine("  -d -data");
            Console.WriteLine("     Path to the texture metadata map (TEXTURES only).");
            Console.WriteLine("  -h -? -help");
            Console.WriteLine("     Prints this message.");
        }
    
        private static List<string> GetDataFiles(string directory, string extension) {
            List<string> files = new List<string>();
            try {
                foreach(string f in Directory.GetFiles(directory)) {
                    if(f.EndsWith(extension))
                        files.Add(f);
                }

                foreach(string d in Directory.GetDirectories(directory))
                    files.AddRange(GetDataFiles(d, extension));
            } catch(System.Exception except) {
                Console.WriteLine("Folder filter error!\n" + except.ToString());
            } 
            return files;
        }
    
        #region "Textures"

        private static Dictionary<string, TextureMetadata> GetTextureData(Options option) {
            Dictionary<string, TextureMetadata> data = new Dictionary<string, TextureMetadata>();
            List<string> files = GetDataFiles(option.input_dir, option.extension);

            try {
                string contentsMaps = File.ReadAllText(option.input_meta);
                List<string> mapData = new List<string>(contentsMaps.Split('\n'));
                mapData.ForEach(d => {
                    string[] meta = d.Split(' ');
                    if(files.Contains(meta[0])) {
                        string id = meta[0].Remove(0, option.input_dir.Length).Split('.')[0];
                        if(id.StartsWith("/"))
                            id = id.Remove(0, 1);
                        TextureMetadata tex = new TextureMetadata();
                        tex.populate(meta, id);
                        data.Add(meta[0], tex);
                    }
                });

                return data;
            } catch(System.Exception except) {
                Console.WriteLine("Texheader creation error!\n" + except.ToString());
            }

            return null;
        }

        private static List<DataEntry> CreateBlobDataTexture(Dictionary<string, TextureMetadata> data, string output) {
            using(MemoryStream stream = new MemoryStream()) {
                using(BinaryWriter writer = new BinaryWriter(stream)) {
                    List<DataEntry> dataEntries = new List<DataEntry>();

                    foreach(KeyValuePair<string, TextureMetadata> entry in data) {
                        DataEntry info = new DataEntry();
                        info.data_id = entry.Value.fileName;
                        byte[] header = entry.Value.createData();
                        info.start = (uint)stream.Length;
                        writer.Write(header);
                        info.dataStart = (uint)stream.Length;
                        try{
                            byte[] data_buffer = File.ReadAllBytes(entry.Key);
                            writer.Write(data_buffer);
                        }catch(System.Exception except) {
                            Console.WriteLine("Failed to read datafile! " + except.ToString());
                        }

                        info.end = (uint)stream.Length;
                        dataEntries.Add(info);
                    }

                    using(BinaryWriter outstream = new BinaryWriter(new FileStream(output, FileMode.Create))) {
                        try {
                            outstream.Write(stream.ToArray());
                        } catch(System.Exception except) {
                            Console.WriteLine("Failed to write blob! " + except.ToString());
                        }
                    }

                    return dataEntries;
                }
            }
        }

        private static void CreateLookupTableTexture(List<DataEntry> data, string map) {
            List<string> texture_template = new List<string>{
                "#include <textures.h>\n",
                "const size_t TEXTURE_COUNT = {0};\n",
                "data_info TEXTURES[] = {",
                "};\n"
            };

            string centryLine = "   {{ \"{0}\", {1}, {2}, {3}, NULL }},"; 

            using(StreamWriter writer = new StreamWriter(new FileStream(map, FileMode.Create)))  {
                try {
                    writer.WriteLine(texture_template[0]);
                    writer.WriteLine(String.Format(texture_template[1], data.Count));
                    writer.WriteLine(texture_template[2]);

                    data.ForEach(l => writer.WriteLine(String.Format(centryLine, l.data_id, l.start, l.dataStart, l.end)));

                    writer.WriteLine(texture_template[3]);
                } catch(System.Exception except) {
                    Console.WriteLine("Failed writing C Map! " + except.ToString());
                }
            }
        }

        #endregion
    }
}
